add NUTS sampler that behaves mostly like Stan

perrydv opened the pull request:
This PR contains `sampler_NUTS`, which behaves mostly like the sampler rstan runs by default. This has diagonal mass matrix, adaptation scheme choices, and other details.

Some items to do:
- roxygen
- behavior when number of warmup samples is too small
- run on existing and/or new tests
- consider additional control parameters (whether to do adaptation at all, possibly separately for stepsize vs mass matrix, and separate setting of initial epsilon from whether to do stepsize adaptation).

This PR is branched off of master, and at the same time branch add-calcLogProb is for revising the first version of HMC we released.

danielturek commented:
@perrydv @paciorek Correction, iall the various flags for controlling adaptation / initialization are in place. (boolean control flags `adaptive`, `adaptEpsilon`, `adaptM`, `initialEpsilon`)

So, do we want any additional checking of `nwarmup`, aside from the current warning if `nwarmup < 20`?

Otherwise, I guess I'm asking for another set of eyes on all of this, and comments on "what remains to be done" ?

paciorek commented:
Regarding `nwarmup`, I think we definitely want to allow it to be zero. 

As far as `20` as triggering a warning - was that just out of the blue or based on when it becomes infeasible to divide up the warmup length into the various stages? I suppose we could look and see what Stan does, but I'm ok with leaving as is provided we add the `0` option.

As far as warnings, I think we want to warn if (1) `adaptive=FALSE` or (2) if `adaptive=TRUE` and `nwarmup` is 0, as well as if `nwarmup` small.

If a user sets `adaptive=FALSE` then we may not want to warn explicitly about a small warmup size (rather we warn about lack of adaptation) since then we are just in a standard burn-in type situation and it's up to the user to decide what they want in terms of discarding initial iterations.

paciorek commented:
I see this check: `epsilon == 0`. Shouldn't we be checking for smaller than some very small value?  I suppose if we keep taking one-half of the current epsilon a huge number of times we eventually will underflow and see an exact 0. But I would expect that to take a very long time.

paciorek commented:
I think we probably need some revised text in the nimble manual regarding HMC. @danielturek were you planning to do that?

I can take a look at the revised `nimbleHMC` roxygen early next week.

Apart from that, I'm not sure if there's something else I should spend time on given both of you have invested a bunch of time in the actual algorithmic implementation. Thoughts? I have not looked through the new code apart from a very quick look at when adaptation is triggered.

danielturek commented:
Removed (what I feel was an unnecessary) `scalarComponents = TRUE` in the `configureHMC` method.

@paciorek This will hopefully fix the expansion that you noted, but we'll see if testing shows any problem with this change.

danielturek commented:
@paciorek I'm behind on your comments above here (3 comments from August 12), these all look worthwhile to consider and discuss, but I won't get to it now.

To address one thing, I have not done anything really with the User Manual, if anyone wants to make a first go at updating that (w.r.t. HMC) then please go ahead, and I'll certainly review it.  Otherwise, I'll aim to do it sometime down the road.

paciorek commented:
Just to pile a few more things on, this is what I noticed in reviewing the roxygen (including the nimble manual comment I made earlier, for completeness):

- Need to update DESCRIPTION:description which simply refers to Hoffman and Gelman now.
- Revise nimble manual to reflect new NUTS sampler.
- Example code for various nimbleHMC functions has `nimbleOptions(enableDerivs = TRUE)`, but this is the default so I think we should take it out for conciseness and to avoid confusion.
- I don't think it's clear from roxygen that for {add,build,configure,nimble}HMC that `NUTS` is the default. (In fact `type` doesn't have an explicit default, unlike `addSampler` -- probably fine but noting the inconsistency.)

danielturek commented:
@paciorek 

I've cleaned up the warning messages for the `NUTS` sampler, and yes, it does permit `nwarmup = 0` if you set that value manually.

The warning messages regarding small values of `nwarmup` are now only printed if `adaptive = TRUE`.  When `adaptive = FALSE`, these warnings are not printed, but (contrary to previous suggestion) I did not print any warning message saying "warning, NUTS sampler is not doing any adaptation".  My logic is that `adaptive` will only be `FALSE` if a user explicitly sets it that way, and we don't print any sorts of warnings in other (adaptive) samplers when `adaptive = FALSE`.  However, I'm happy to discuss this decision further if anyone wants.

In the check regarding if adaptation of `epsilon` diminishes to `if(epsilon == 0)`, I changed this check to instead be `if(epsilon < 1e-16)`.  That value is chosen as being on par with `.Machine$double.eps`.

I've updated the package `DESCRIPTION` to briefly introduce both the `NUTS` and `NUTS_classic` samplers.  I added a reference for the `NUTS` sampler pointing to version 2.32.2 of Stan (Stan Development Team, 2023) in the `DESCRIPTION`. 
 Noting in advance that I think CRAN will push back on this reference to (Stan Development Team, 2023), and will want a DOI or arXiv number for it, but we'll cross that bridge when we get there.

I removed `nimbleOptions(enableDerivs = TRUE)` from the HMC examples.

I reviewed the argument passing between these functions.  There was actually a reason for the way it was done before (relating to evaluation of named arguments / unnamed arguments), but the changes I just made should function the same.  Anyway, the Roxygen now indicates for each function that "The default sampler type is NUTS", and an explicit default value for the `type` argument of each function is now provided (with default value of `"NUTS"`).

Still to be addressed: User Manual.

danielturek commented:
@paciorek @perrydv I've reviewed and updated the User Manual w.r.t. `NUTS` and `NUTS_classic` samplers.

Any additional review, suggestions, or feedback is welcomed.  Short of that, I'll consider this PR ready to merge (but I welcome additional feedback/suggestions).

perrydv commented:
@paciorek @danielturek 

I am pushing a bunch of changes:

- Make both NUTS and NUTS_classic have the same control args for epsilon and initializeEpsilon. Previously NUTS_classic took an initialEpsilon that, if not provided, resulted in 1. The default is still 1, but having an epsilon and an initializeEpsilon argument provides finer control (can provide a starting value and also have it go through the initEpsilon procedure or not). I also think it would be confusing and seem arbitrary if the two samplers have different control args for corresponding features.
- In NUTS_classic, make the first mu be set *before* the first initEpsilon step. This seems to be how the Stan source code works, and therefore how our NUTS works.
- In NUTS_classic, make the handling of too-small nwarmup values match NUTS, which matches Stan (for both NUTS and NUTS classic). This puts (rounded) 15% / 75% / 10% in the initial buffer / sampling for adaptation / terminal buffer.
- For the issue of nwarmup < 20, Stan simply errors out. We want to allow nwarmup==0 as a way to skip warmup. (That is somewhat redundant because we also have an `adaptive` parameter that if `FALSE` indicates no adaptation, but I still like the concept of supporting nwarmup==0).  So what I have done is error out for 1 <= nwarmup <= 19. This means you can deliberately request 0 warmup, but if you want warmup it must be >= 20 to be meaningful. I'm just not sure what the lower bound would be. 
- Again to match NUTS and NUTS_classic in terms of user interface (control args), provided both `adaptEpsilon` and `adaptM` args (as well as the `adaptive`, which if FALSE makes the other two irrelevant). These provide finer control over what is adapted.
- Importantly, made the call to initEpsilon in NUTS_classic after completing a window of var (M) adaptation *NOT* reset epsilon to 1 before doing the initialization sequence. It does get re-initialized, but starting from its previous value.
- Added a test-tuning.R file. This contains tests of the var (M) tuning windows and the control args `adaptive`, `initializeEpsilon`, `epsilon`, `adaptM`, and `adaptEpsilon`.

Here are the remaining issues I see:
- In NUTS_classic, there is ad hoc handling of cases where the p or q have any NaNs or NAs. It is to slash epsilon and try again. In our NUTS, following Stan, invalid values result in bailing out of the tree. Can we do that in our NUTS_classic too?
- It seems we can only set nwarmup as a control arg, and then burnin is ignored. But it looks like burnin is passed to `before_chain` and could be used there. Given that burnin is another name for warmup and that run-time args provide greater control than build-time args, should we instead use the burnin information provided when running the MCMC? I can see that could be an issue if the NUTS wants to use the default of half the samples regardless, but is there another way we can provide control? Perhaps a control arg saying whether to respect the burnin argument or not? Otherwise we are *forcing a re-build and re-compilation* for non-expert users just wanting to try different nwarmup.  (An expert user can work out how to set it but it's not easy.)

paciorek commented:
@perrydv please see my issue #35 regarding how I was understanding the difference between `nburnin` and `nwarmup` and the potential difficulty that entails. 

We might be ok thinking of 'burnin' as "how many samples to discard" and 'warmup' as specifically pertaining to HMC adaptation, though I think Gelman et al. want to use "warmup" to replace "burnin" in terms of terminology. I think that is basically how things are implemented now in nimble/nimbleHMC, but open question is whether we're ok with it and whether it will confuse users that we throw out samples based on `nburnin` not based on `nwarmup`. I suppose we could have `nburnin` in the HMC case control how many post-warmup iterations to throw out and always throw out the warmup ones. I do like users having the option to keep the warmup samples just in case they are curious to look at them.

perrydv commented:
Fixing test failures:

- I fixed infrastructure of one of the new test-tuning I had introduced. (These test code involving the tuning, not performance of the tuning.)

- @danielturek I changed the Dirichlet-multinomial test so that the anticipated posterior mean of `p` will be `y/n`, rather than the `p` used to simulate the single (vector) datum `y`. I believe that as long as the prior on `alpha` is not having much influence, the estimated proportions should be the sample proportions, not the (unknown) true proportions. The test was failing for this reason. I was able to use a much smaller threshold once this change was made.

- @danielturek There seems to be a glitch in `configureHMC` where `includeConfGetUnsampledNodes` is used but unknown. Does this sound related to any changes you made?

perrydv commented:
@paciorek Thanks for pointing out issue #35. You emphasized the important point that by providing two separate numbers (nwarmup and nburnin) we are allowing users to investigate warmup samples if they want. I was looking at a different use case, where in test code (I think) we have niter = 20000 and nburnin = 10000.  The NUTS tuning will go by default for half of niter or for a specific number that was set, so in this case nburnin might match nwarmup. But I'm worried a user might think they are choosing nwarmup when they provide nburnin at *runtime*, not realizing that any *warmup* settings were locked in at *build time*.

Would we want to:

- add another control parameter saying whether to use nburnin as nwarmup, defaulting to TRUE? (This expands the control parameters, but arguably once a user gets digging into them it is because they want fine-grained control so they might appreciated being given fine-grained control.)
- somehow provide post-build capability to change control parameters, although I'm not sure we have a great way to do that.

perrydv commented:
In terms of the test failures, I am getting confused because I am seeing different results on the test machines than my machine. The tests pass on my machine except for a glitch in configureHMC. I don't get the outside-tolerance failures shown. Is anyone else able to run through the tests? Not sure it's helpful for me to get stuck in system-specific issues.

paciorek commented:
I think allowing a user to say that `nwarmup` should equal `nburnin` to avoid lock-in at build time make sense. I guess a further question is whether the default for `nwarmup` should be `nburnin` and finer grain control comes from a user saying they want to separate the concepts and providing an explicit `nwarmup`. Then we also have to decide what to do if `nburnin` is zero. Should that cause no adaptation or should it result in half the sames being used for warmup...

perrydv commented:
# On nwarmup vs nburnin

@paciorek  My instinct was what you said: default behavior is that nwarmup and nburnin are synonyms, and the user must say so at build time if they want to separate them. 

Yes the `nburnin=0` (default) case also struck me as a potential confusion after I commented yesterday. One point is that if they want no adaptation, they can set `adaptive=FALSE`. This starts to get a little complicated but we could say:

- if nwarmup==-1, then use nwarmup=nburnin if nburnin > 0 and use nwarmup = niter/2 if nburnin==0 and adaptive==TRUE. But if adaptive==FALSE use nwarmup==0 regardless.
- if nwarmup==-2,  then use nwarmup=nburnin if adaptive==TRUE, even if nburnin ==0. 

In other words, nwarmup==-1 (default) respects non-zero nburnin and otherwise defaults to half niter.  nwarmup==-2 respects nburnin no matter what and never defaults to half niter.  And nwarmup is always disabled by adaptive=FALSE.

A potentially missing piece of the system is that there is no way (I don't think) for the NUTS sampler to know if it is the only sampler on the model or not.  If there are other samplers, that could be a source of confusion about adaptation phases.

# On testing:

- I changed a tolerance on the Dirichlet-multinomial test because I had only checked it for one of the runs, not both.
- I changed `conf$show(...)` to `conf$show()` because it looks like an MCMC configuration object's `show` method does not take arguments and when `...` was non-empty, it caused an error.
- There was a glitches from the non-standard evaluation of `test_that` in one of the new tests I wrote and I (again) think I fixed it.

paciorek commented:
The `-1`, `-2` idea for `nwarmup` is feeling overly complicated to me. Does the following satisfy what we need?

- if `nwarmup` is not set, then set `nwarmup` to `nburnin`. If `nburnin=0`  then no warmup/no adaptation.
- if `nwarmup` is set to -1, then set `nwarmup` to one-half of `niter`.
- if `nwarmup` is set to a number equal to 20 or greater then use that value.
- if `nwarmup` is set to 0, then no warmup/no adaptation.

On a related note, can we remove `adaptive` and just have it controlled by the number of warmup iterations?

In all cases, I think we should print the number of warmup iterations being used so that it is clear to the user. And we might say something about burnin as well when an HMC sampler is used, indicating how many iterations are being discarded so a user knows if any of the samples they get are still part of warmup.

paciorek commented:
I verified with Paul that `set.seed(1)` seems to produce the same stream of random numbers on Windows as on Linux.

```
> set.seed(1)
> rnorm(1)
[1] -0.6264538
```

danielturek commented:
I've reviewed the changes for the adaptation schedules of the `NUTS` and `NUTS_classic` samplers, including the warmup windows, adaptation of `epsilon`, and control parameters between the two samplers.  The minimum threshold of 20 warmup iterations does seem arbitrary, but I understand that's what Stan does, and it's fine with me.

@perrydv In the new handling of the `epsilon` control argument for the `NUTS_classic` sampler, the current behavior has a default value for `epsilon` of 0, then on the first sampling iteration:
```
if(epsilon == 0) epsilon <<- 1
```
Why not (more simply) have the default value for `epsilon` control argument be 1, as:
```
epsilon <- extractControlElement(control, "epsilon", 1)
```
Same for in the `NUTS` sampler?

@perrydv Do we have any interest in pulling out some of the common functionality between `sampler_NUTS` and `sampler_NUTS_classic` into standalone NFs, which would then be nested within each HMC sampler?  I'm looking right now at the routines for defining the HMC warmup intervals, which takes place in the `before_chain` methods of both samplers.  Pulling this out into a separate (nested) NF has the advantage of cleaner code and being easier to maintain, but the downside of additional nested NFs for compilation.  Given the complexity of the HMC samplers already, I'm somewhat in favor of doing this.  What do you think?

@perrydv You made the comment earlier in this thread that "In NUTS_classic, there is ad hoc handling of cases where the p or q have any NaNs or NAs. It is to slash epsilon and try again. In our NUTS, following Stan, invalid values result in bailing out of the tree. Can we do that in our NUTS_classic too?"  Looking at the code for the `NUTS_classic` sampler, there is code in the `initEpsilon` method, where if p or q becomes NaN, then the value of epsilon is reduced by a factor of 1000.  Is this what you're talking about?   If so, then I think to match the `sampler_NUTS`, this would only change to reduce epsilon by a factor of 2, rather than this extreme factor of 1000, until we get a foothold (a transition to a valid point in parameter space).  (Separately, in the sampling routine (specifically, the `buildtree` method of `sampler_HMC_classic`) when a NaN value of p or q is encountered, we do "bail out" of that search tree, by setting a few control-flow variables to 0.)

@perrydv @paciorek I bumped up the package version number of `nimbleHMC` to 0.2.0.

Yes, the `includeConfGetUnsampledNodes ` argument was added to prevent the issuance of duplicated warnings from the HMC configuration functions.  On the `devel` branch of `nimble`, `conf$show(...)` does take arguments.   I'll leave the `conf$show(...)` out of `nimbleHMC` for now, but we'll go back to having some double-warnings, until next release of `nimble` (and I can add `conf$show(...)` back into `nimbleHMC`).

## burnin vs. warmup debate

Also referencing @paciorek comment #35.

This warmup vs. burnin question is surprisingly difficult.

I also feel `nwarmup` being `-1` or `-2` having different functionality is overly complicated.

@paciorek Made the suggestion that:

- if `nwarmup` is not set, then set `nwarmup` to `nburnin`. If `nburnin = 0` then no warmup/no adaptation.
- if `nwarmup` is set to `-1`, then set `nwarmup` to one-half of niter.
- if `nwarmup` is set to a number equal to `20` or greater then use that value.
- if `nwarmup` is set to `0`, then no warmup/no adaptation.

Isn't "`nwarmup` not set" the same as "`nwarmup` is set to `-1`" ?  How are the first and second bullet points different?

Is there any merit to:
- Keep `nburnin` as the number of MCMC samples to discard.  Nothing more complex. (I would really prefer this).
- `nwarmup` is controlled as present, by 1/2 of `niter`, unless a user specifies otherwise at the time of configuration.
- If a user wants to experiment with different `nwarmup` periods, they always have runMCMC(..., reset = FALSE) available.  And we explain this in documentation?

No solution seems ideal.  I just don't want to in extricably intertwine the (presently *very simple*) concept of `nburnin` with HMC adaptation.

If we really, really want another way to set `nwarmup` at *runtime*, while keeping it independent from `niter`, then we could develop some way for *runtime* information to be passed from `runMCMC` arguments into the samplers.  This already happens with the information of `niter`, `nburnin`, and `nchains` going into the `before_chain` method of samplers.  I just don't want to overload `nburnin`.  So, if we want to do this, then how about be another argument to `runMCMC`, perhaps `auxiliarySamplerInfo` or `auxiliarySamplerData`, which gets passed along into `before_chain` methods, and can manually determine the number of warmup iterations.  That said, my current preference remains the (my) 3 bullet points above.

danielturek commented:
@perrydv @paciorek Two other thoughts I've had on this:

1. If we must determine nwarmp from either niter or nburnin, then fine.  But I don't think it's a good idea to tie it to *both* of them, in some complex manner.

2. If we really want to add additional control over nwarmup at MCMC runtime - which I agree is a nice idea - then how about we (I) add another HMC utility function:
```
setWarmupHMC(obj, nwarmup, nodes)
```
where `obj` is either an MCMC configuration object, or a compiled/uncompiled MCMC, `nwarmup` is the number of warmup iterations to assign, and `nodes` is optional, if you only want to set nwarup for the HMC samplers operating on some `nodes`.

perrydv commented:
@danielturek @paciorek Sorry I wasn't caught up on the past few days of comments when we discussed all of this yesterday. One point from here that we didn't discuss was whether default epsilon of 0, which is then set to 1, should be replaced by having a default of 1. My thought was that it could be useful to know whether they did not set epsilon (so it is 0, and then we use a default of 1) or that they set epsilon to a chosen value, and that value might be 1. The value provided in the control list is retained as epsilonOrig. I'm not sure off the top of my head of a use case where it matters, but I had something in mind at the time, and it does seem to be a bit richer information.

perrydv commented:
Also, I am not opposed to the `setWarmupHMC` idea, but let's defer until a future release. It could perhaps be more generic (finding a changing control args in samplers in general), or that may be more than we want to tackle.

perrydv commented:
I think the Windows vs Mac testing issues have resolved and I still suspect there was a simple infrastructure issue. I have added more to test-tuning, which had not previously caught the problem resulting from a stray `epsilon <- 0` in NUTS_classic setup code. I have now removed that too. Tests seem to be passing.

danielturek commented:
I've updated `sampler_NUTS` and `sampler_NUTS_classic` to implement the new `warmupMode` behavior.  I have not done extensive testing of this (only minimal), and I won't be able to again until this evening (Saturday 9/9).  @paciorek You mentioned you might have some time available.  Here are things which could use attention or additional work:

- Review and critique of the "informative messages" that each HMC sampler now outputs, regarding how `nwarmup` was selected for each of the different cases.  Do they all seem reasonable?
- Review the additions to the Roxygen documentation for the new `warmupMode` and `warmup` control arguments.
- Add any number of tests, to test out the new functionality for selecting `nwarmup` (including all cases when warmupMode is each of "default", "burnin", "fraction", and "iterations".
- Open question per another email: when warmupMode is "default" and nburnin=0, do we want the number of warmup iterations to be niter/2, or min(niter/2, 1000) ?

@perrydv @paciorek One problem I ran into when I quickly tried to do some testing: I found that using the tools I'm aware of for modifying private data in compiled nimbleFunctions - specifically using either:
```
nimbleOptions(buildInterfacesForCompiledNestedNimbleFunctions = TRUE))
```
or 
```
valueInCompiledNimbleFunction(...)
```
I was unable to modify the `warmupMode` private data value.  This could make sense if it doesn't persist as private data because it's never used in compiled methods; or could make sense if these facilities were never developed for character strings.  Running out of time, I did not investigate further.

danielturek commented:
@paciorek @perrydv I did not adapt any of the tests in the new test file `test-tuning.R`, which on quick glance use the old system of `control = list(nwarmup = 1000)` to set a warmup period of 1000.  This would need to be replaced by `control = list(warmupMode = "iterations", warmup = 1000)`.

danielturek commented:
Realizing also my earlier attempt at changing private data `warmupMode` is nonsensical, since this gets baked into a nested nimbleFunction.  I do hope we keep this implementation, though, to avoid completely duplicating the setting-warmup related code.

paciorek commented:
Thanks @danielturek . I will try to do a bit now-ish and then I will make some time tomorrow.

paciorek commented:
Ok, I made some progress. About to commit by 4 pm PT.

1. I added tests of the various `warmupMode` values.
2. If a user specifies "burnin" with `nburnin=0`, I added a special case message of "is using 0 warmup iterations, so no adaptation is being done, apart from initialization of epsilon (if `initializeEpsilon` is TRUE).". LMK if I got this wrong (or just fix it).
3. I made some minor edits to roxygen (but have not rebuilt the Rd files) and to the burnin/warmup output messages. 
4. I split the long messages about warmup/burnin into multiple lines as we generally do. May still want to clean that up a bit.

Some musing about the messages about nwarmup:

1. I'm wondering if we should actually do the arithmetic to say exactly how many pre- and post-warmup samples are returned. This will require getting `thin`. I suppose a complication occurs if a user has `thin2`. So maybe too complicated, at least for now.
2. My vote would be to make the messages less verbose and not do all the explanation of why the nwarmup is as it is, but merely say what it is and then say if any of the returned samples are from warmup. But that's just my thought for now. And happy to leave as is for this release.

paciorek commented:
I made edits to `sampler_NUTS`. I hope that is right given there is now `sampler_NUTS_rev2` branch.

danielturek commented:
@paciorek you really combed over this.  Thank you.  The revised HMC messages all look fine.

I updated the existing tests `test-tuning.R`.  A small number of the tests which modified the warmup-related data in the sampler functions, I commented out.  @perrydv If you think these tests are critical, it would require a little more care to update these tests.

I'd vote against doing the arithmetic of the samples returned, for one the `thin` issue you mentioned, and for two it seems a little (more) overboard than present messages.

Tests are now passing.  I'm fine with this as currently is.  I'll wait for final review and OK before going ahead with submission.

paciorek commented:
Great, thanks. I'll take a quick look this afternoon.

@perrydv did you want to take a look? If so, any chance you'd have time today so we can push this out as soon as Daniel has a chance to?

danielturek commented:
Yes, I'll wait to hear from @perrydv for any final review or comments.

perrydv commented:
LGTM. @danielturek I updated the last two tests of test-tuning and put them back in action.  I am not sure how general it is but we seem to get an error upon building a second HMC sampler and compiling it for the same model as the first HMC sampler. It has to do with modelValues definitions. We could fix that in a later release.

danielturek commented:
Tests passing, I'm merging this into `master` and proceeding with CRAN submission
